Add text translation capabilities

mklewandowski opened the pull request:
This PR adds a text translation mechanism to the starter-projects.  It is heavily based upon the CFM translation model seen here:
https://github.com/concord-consortium/cloud-file-manager/blob/master/src/code/utils/translate.js

I tested this in Chrome, Firefox, Edge, Safari.  Works well in all cases by both changing the lang property in `index.html` or by changing the browser settings.

Notes on implementation:
- translation json files are added to `utils/lang`
- translation is handled in `utils/lang/translate.ts` and can be added by importing the `translate` function and calling the function with the appropriate translation key:
`t("INTRO.HELLO")`
- ~~I'm not crazy about the translation keys using a tilde, but this was the approach used in the CFM, and I wanted to be consistent. Also it makes it much easier to search for text keys.  Does anyone have an opinion on this?~~
**EDIT**: I changed the code and removed the tilde
- by default we choose the language by first looking at the HTML DOM lang property of the root element of the document and then by looking at first valid language specified by the browser.  This is the technique used by the CFM, and I think it is fine as long as we know to remove/update/change the `lang="en"` attribute from `index.html`.  I'm also open to switching the check and looking at the browser specified language *first*.  
**EDIT**: I think we should leave the checks as is but remove the lang property in `index.html` so that the translation works out of the box by changing browser settings.
- ~~I made the decision to remove a couple of the optional parameters (`vars`, `lang`) that were used in the CFM `translate` function seen here:
https://github.com/concord-consortium/cloud-file-manager/blob/master/src/code/utils/translate.js#L77
I'm open to bringing these params back, but it seemed that they added extra complexity to a function that at least for a base implementation should be fairly simple.  Thoughts?~~
**EDIT**: I added all of the parameters seen in the CFM function.
- ~~I chose not to translate the intro "Hello World" text that is first displayed in the app.  I don't want users of the starter-projects to feel as if the translation is a tool that *needs* to be used, but rather a lightweight option that is there by default which can be used, left for future use, or removed.~~
**EDIT**: the intro hello world text is now wrapped in the translation function.  Users of the starter-projects can remove the translation if they don't want it, but it is in use by default.

kswenson commented:
Response to notes on implementation (prior to review):

>I'm not crazy about the translation keys using a tilde, but this was the approach used in the CFM, and I wanted to be consistent. Also it makes it much easier to search for text keys. Does anyone have an opinion on this?

I don't recall if there was any particular reason for the tilde (I think it predated my involvement). I'd be fine with eliminating it and just using the uppercase constants.

>by default we choose the language by first looking at the HTML DOM lang property of the root element of the document and then by looking at first valid language specified by the browser. This is the technique used by the CFM, and I think it is fine as long as we know to remove/update/change the lang="en" attribute from index.html (I added a comment in translate.ts about this). I'm also open to switching the check and looking at the browser specified language first.

I think this may have been a response to use with CODAP/SproutCore which handles localization by generating separate language-coded static HTML pages for each language. In that context, once you've been directed to a static HTML page coded to a particular language it's better to be consistent. If our HTML pages aren't language-coded then it shouldn't matter. If we believe that our HTML pages should be language-coded but then the browser language should override the page we can change it, but I suspect it won't really matter.

>I made the decision to remove a couple of the optional parameters (vars, lang) that were used in the CFM translate function seen here:
https://github.com/concord-consortium/cloud-file-manager/blob/master/src/code/utils/translate.js#L77
I'm open to bringing these params back, but it seemed that they added extra complexity to a function that at least for a base implementation should be fairly simple. Thoughts?

Variable replacement is critical. You can't do localization without it. (`You have 2 new messages.`) I don't recall the rationale for allowing the client to override the language, but maybe it had something to do with explicit language references, e.g. `English: "foo", Spanish: "bar", Your Language: "baz"`. I'd be inclined to take the function as it's used in CFM/CODAP/Sage since that's where we've made the heaviest use of translation. Presumably, the features are there because they came up in real use cases.

>I chose not to translate the intro "Hello World" text that is first displayed in the app. I don't want users of the starter-projects to feel as if the translation is a tool that needs to be used, but rather a lightweight option that is there by default which can be used, left for future use, or removed.

I understand that rationale, but I think we should make it easy to do the "right" thing while still allowing the expedient thing. I don't think it would be a bad thing if by default CC apps were set up to support localization. If someone's in a hurry and doesn't want to deal with it they can just use strings instead of `t(STRING_KEY)`.

kswenson commented on src/utils/translation/translate.ts:
```diff
@@ -0,0 +1,55 @@
+// To use, remove lang attribute from index.html, import translate function, and call with translation key:
+// translate("~INTRO.HELLO")
```
Conventional usage should look like this:
```typescript
import t from "utils/translate";

console.log(t("~INTRO.HELLO"));
```
The call to the translate function should be as inconspicuous as possible.

mklewandowski replied:
Yeah, that's how I was using it in practice, but you're right that this comment makes it unclear.  I'll clarify so it makes more sense.

kswenson replied:
Comment could still use some improving...

mklewandowski replied:
Fixed.

kswenson commented on src/utils/translation/translate.ts:
```diff
@@ -0,0 +1,55 @@
+
+const translate = (key: string) => {
+  return translations[defaultLang][key] || `UNKNOWN KEY: ${key}`;
+};
```
As mentioned in the comments, I think you should bring over the full `translate()` function. Variable replacement is required, and if CODAP/CFM found a use for the additional `lang` parameter I'd be willing to bet there was a good reason for it.

mklewandowski replied:
I added these parameters and the additional functionality to the translate function.  No one needs to use them so I don't think they get in the way at all of a simple implementation.

kswenson replied:
Thinking about this more, and following up on @scytacki's suggestion of thinking of this as the beginnings of a more widely used library within CC, I think we should change the API of the `translate()` function to be more extensible down the road. Something like the following:
```typescript
interface ITranslateOptions {
  lang?: string;
  count?: number;
  vars?: Record<string, string>;
}
export default function translate(str: string, options?: ITranslateOptions {
  ...
}
```

mklewandowski replied:
Looks good.  I added these changes.

kswenson reviewed (changes requested):
As mentioned elsewhere, my vote would be to:
1. bring over the full `translate()` function with support for all arguments.
2. use translation by default in the starter project.

but others may feel differently.

mklewandowski commented:
> I think this may have been a response to use with CODAP/SproutCore which handles localization by generating separate language-coded static HTML pages for each language. In that context, once you've been directed to a static HTML page coded to a particular language it's better to be consistent. If our HTML pages aren't language-coded then it shouldn't matter. If we believe that our HTML pages should be language-coded but then the browser language should override the page we can change it, but I suspect it won't really matter.

@kswenson at present, the starter-projects does have a static language set in the HTML page (`en`).  If we're going to turn translation on by default then we should probably remove the static language setting (so we can actually see the translation working.  I think it's still fine to leave in the `getPageLanguage` test and to have it be the first thing checked to determine the language, but no language should on the page be set by default (let the engineers add it if needed).  Then we can actually see the translation working by changing the browser language.

mklewandowski commented:
@kswenson as per your feedback I did the following:
- added parameters from the CFM example back into the translate function
- turned on translation by default ("Hello World" has Spanish translation available).  In line with this, the lang property in index.html has been removed.  Users can see the translation in action by running the starter project and changing the browser setting to Spanish.

kswenson commented on src/utils/translation/translate.ts:
```diff
@@ -0,0 +1,64 @@
+                      ? baseLang
+                      : "en";
+
+const translate = (key: string, vars: any = {}, lang: string = defaultLang) => {
```
```suggestion
const translate = (key: string, vars: Record<string, string>, lang = defaultLang) => {
```

mklewandowski replied:
Ah, thanks!

kswenson reviewed (approved):
👍 Looks good -- still one comment to be improved and one suggestion to improve types but I trust you to take it from here. 😉

scytacki reviewed (approved):
It be nice to include some doc on how to use variables in the strings. That could just be a comment in the translate.ts file.
It'd also be nice to publish this is a package so it is easier to update after someone has cloned the starter project. 
But those suggestions shouldn't hold up the PR.

kswenson reviewed (changes requested):
I added a comment about changing the API of the `translate()` function that will be easier to do now than later.

mklewandowski commented:
> It be nice to include some doc on how to use variables in the strings. That could just be a comment in the translate.ts file.
> It'd also be nice to publish this is a package so it is easier to update after someone has cloned the starter project.
> But those suggestions shouldn't hold up the PR.

I added a README file that goes into more detail and shows some examples.  Should get us closer to eventually carving this out into an independent translation library/tool.

mklewandowski commented:
> I added a comment about changing the API of the `translate()` function that will be easier to do now than later.

I like your ideas.  I updated the translate function.

kswenson reviewed (approved):
👍 Looks great! I made a few suggestions on some of the comments. For reference:
- Internationalization: Making your application localizable.
- Localization: Targeting your application to a particular locale, including simple translation but also potentially including more nuanced things like pluralization, date-formatting, currency-formatting, etc.
- Translation: Simple one-to-one string conversion.

My point in some of the suggested comment changes is that we should be thinking of this effort in general, and the `translate()` function in particular, as part of a larger localization effort rather than as just a simple string translation mechanism.

mklewandowski commented on src/utils/translation/README.md:
```diff
@@ -0,0 +1,45 @@
+# Starter Projects Translation Functions
+
+The modules within `utils/translation` can be used to add text translation to a starter-projects based application.
```
Fixed.

mklewandowski commented on src/utils/translation/README.md:
````diff
@@ -0,0 +1,45 @@
+}
+```
+
+#### Using the translation function
````
Fixed.

mklewandowski commented on src/utils/translation/README.md:
```diff
@@ -0,0 +1,45 @@
+
+
+#### Determining the current language
+The `translate` function will first get the HTML DOM `lang` attribute of the root element of the document to determine the current page language.  If no `lang` attribute is specified, then the `translate` function will get the first valid language specified by the browser to detgermine the current page language.  Optionally, a language value can be specified when calling the `translate` function  ...
```
Fixed.

mklewandowski commented on src/utils/translation/README.md:
```diff
@@ -0,0 +1,45 @@
+# Starter Projects Translation Functions
```
Fixed.

mklewandowski commented:
@kswenson thanks for the readme suggestions.  I made changes and think this is finally ready to merge.
